modules/imgproc/src/color_lab.cpp

```diff
@@ -1563,96 +1357,93 @@ static inline void trilinearInterpolate(int cx, int cy, int cz, const int16_t* L
-    v_store_aligned(addrofs + 4, addrDw1);
-
-    const int16_t* ptr;
-#define LOAD_ABC(n) ptr = LUT + addrofs[n]; a##n = v_load(ptr); b##n = v_load(ptr + 8); c##n = v_load(ptr + 16)
-    LOAD_ABC(0);
-    LOAD_ABC(1);
-    LOAD_ABC(2);
-    LOAD_ABC(3);
-    LOAD_ABC(4);
-    LOAD_ABC(5);
-    LOAD_ABC(6);
-    LOAD_ABC(7);
-#undef LOAD_ABC
-
-    //interpolation weights for pix0, pix1, .., pix7
-    v_int16x8 w0, w1, w2, w3, w4, w5, w6, w7;
-    v_mul_expand(v_setall_u16(8), fracX, addrDw0, addrDw1);
-    v_mul_expand(v_setall_u16(8*TRILINEAR_BASE), fracY, addrDw10, addrDw11);
-    addrDw0 += addrDw10; addrDw1 += addrDw11;
-    v_mul_expand(v_setall_u16(8*TRILINEAR_BASE*TRILINEAR_BASE), fracZ, addrDw10, addrDw11);
-    addrDw0 += addrDw10; addrDw1 += addrDw11;
-
-    v_store_aligned(addrofs, addrDw0);
-    v_store_aligned(addrofs + 4, addrDw1);
-
-#define LOAD_W(n) ptr = trilinearLUT + addrofs[n]; w##n = v_load(ptr)
-    LOAD_W(0);
-    LOAD_W(1);
-    LOAD_W(2);
-    LOAD_W(3);
-    LOAD_W(4);
-    LOAD_W(5);
-    LOAD_W(6);
-    LOAD_W(7);
-#undef LOAD_W
-
-    //outA = descale(v_reg<8>(sum(dot(ai, wi))))
-    v_uint32x4 part0, part1;
-#define DOT_SHIFT_PACK(l, ll) \
-    part0 = v_uint32x4(v_reduce_sum(v_dotprod(l##0, w0)),\
-                       v_reduce_sum(v_dotprod(l##1, w1)),\
-                       v_reduce_sum(v_dotprod(l##2, w2)),\
-                       v_reduce_sum(v_dotprod(l##3, w3)));\
-    part1 = v_uint32x4(v_reduce_sum(v_dotprod(l##4, w4)),\
-                       v_reduce_sum(v_dotprod(l##5, w5)),\
-                       v_reduce_sum(v_dotprod(l##6, w6)),\
-                       v_reduce_sum(v_dotprod(l##7, w7)));\
-    (ll) = v_rshr_pack<trilinear_shift*3>(part0, part1)
-
-    DOT_SHIFT_PACK(a, outA);
-    DOT_SHIFT_PACK(b, outB);
-    DOT_SHIFT_PACK(c, outC);
-
-#undef DOT_SHIFT_PACK
+    v_uint16 bitMaskReg = vx_setall_u16(bitMask);
+    v_uint16 fracX = (inX >> (lab_base_shift - 8 - 1)) & bitMaskReg;
+    v_uint16 fracY = (inY >> (lab_base_shift - 8 - 1)) & bitMaskReg;
+    v_uint16 fracZ = (inZ >> (lab_base_shift - 8 - 1)) & bitMaskReg;
+
+    // trilinearIdx = 8*x + 8*TRILINEAR_BASE*y + 8*TRILINEAR_BASE*TRILINEAR_BASE*z
+    v_uint32 trilinearIdx0, trilinearIdx1;
+    v_uint32 fracX0, fracX1, fracY0, fracY1, fracZ0, fracZ1;
+    v_expand(fracX, fracX0, fracX1);
+    v_expand(fracY, fracY0, fracY1);
+    v_expand(fracZ, fracZ0, fracZ1);
+
+    trilinearIdx0 = (fracX0 << 3) + (fracY0 << (3+trilinear_shift)) + (fracZ0 << (3+trilinear_shift*2));
+    trilinearIdx1 = (fracX1 << 3) + (fracY1 << (3+trilinear_shift)) + (fracZ1 << (3+trilinear_shift*2));
+
+    uint32_t CV_DECL_ALIGNED(MAX_ALIGN) vtrilinearIdx[vsize];
+    v_store_aligned(vtrilinearIdx + 0*vsize/2, trilinearIdx0);
+    v_store_aligned(vtrilinearIdx + 1*vsize/2, trilinearIdx1);
+
+    v_uint32 a0, a1, b0, b1, c0, c1;
+
+    uint32_t CV_DECL_ALIGNED(MAX_ALIGN) va[vsize], vb[vsize], vc[vsize];
+    for(int j = 0; j < vsize; j++)
```

Comment:
This implementation is not efficient in comparison with previous one.
Specialization for SIMD128 should be kept back ( https://github.com/alalek/opencv/commit/pr14106_r )